filemanager/s3manager_v2.go

```diff
@@ -0,0 +1,366 @@
+package filemanager
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	s3manager "github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/types"
+	"github.com/aws/smithy-go"
+	"github.com/mitchellh/mapstructure"
+	"github.com/samber/lo"
+
+	awsutil "github.com/rudderlabs/rudder-go-kit/awsutil_v2"
+	appConfig "github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+)
+
+const ServiceName = "s3"
+
+// S3ManagerV2 manages S3 file operations using AWS SDK v2.
+type S3ManagerV2 struct {
+	*baseManager
+	config *S3Config
+
+	sessionConfig *awsutil.SessionConfig
+	client        *s3.Client
+	clientMu      sync.Mutex
+}
+
+// NewS3ManagerV2 creates a new file manager for S3 using v2 AWS SDK.
+func NewS3ManagerV2(
+	appConfig *appConfig.Config, config map[string]interface{}, log logger.Logger, defaultTimeout func() time.Duration,
+) (*S3ManagerV2, error) {
+	var s3Config S3Config
+	if err := mapstructure.Decode(config, &s3Config); err != nil {
+		return nil, fmt.Errorf("failed to decode S3 config: %w", err)
+	}
+
+	sessionConfig, err := awsutil.NewSimpleSessionConfig(config, ServiceName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create AWS session config: %w", err)
+	}
+
+	s3Config.RegionHint = appConfig.GetString("AWS_S3_REGION_HINT", "us-east-1")
```

Comment:
how can we use our app config, in case we're accessing a customer's s3?

Reply:
We have Region and RegionHint in the S3 config. Region refers to the one we get from customers config